packages/postgres-database/src/simcore_postgres_database/utils_folders_v2.py

```diff
@@ -0,0 +1,916 @@
+
+
+def _get_all_permissions(permissions: _FolderPermissions, table) -> ColumnElement:
+    return sa.and_(
+        table.c.read.is_(permissions["read"]),
+        table.c.write.is_(permissions["write"]),
+        table.c.delete.is_(permissions["delete"]),
+    )
+
+
+###
+### UTILS NAMING
+###
+
+_RE_FOLDER_NAME_INVALID_CHARS: Final[str] = r'[<>:"/\\|?*]'
+_FOLDER_NAMES_RESERVED_WINDOWS: Final[set[str]] = {
+    "CON",
+    "PRN",
+    "AUX",
+    "NUL",
+    *[f"COM{i}" for i in range(1, 10)],
+    *[f"LPT{i}" for i in range(1, 10)],
+}
+_FOLDER_NAME_MAX_LENGTH: Final[NonNegativeInt] = 255
+
+
+def _validate_folder_name(value: str) -> None:
+    if not value:
+        raise InvalidFolderNameError(name=value, reason="folder cannot be empty")
+
+    if re.search(_RE_FOLDER_NAME_INVALID_CHARS, value):
+        reason = f"name contains invalid characters. Must comply to regex={_RE_FOLDER_NAME_INVALID_CHARS}"
+        raise InvalidFolderNameError(name=value, reason=reason)
+
+    if value.upper() in _FOLDER_NAMES_RESERVED_WINDOWS:
+        raise InvalidFolderNameError(
+            name=value,
+            reason=f"name is a reserved word in Windows. Can't be any of the following: {_FOLDER_NAMES_RESERVED_WINDOWS}",
+        )
+
+    if len(value) > _FOLDER_NAME_MAX_LENGTH:
+        raise InvalidFolderNameError(
+            name=value,
+            reason=f"name is too long. Maximum length is {_FOLDER_NAME_MAX_LENGTH} characters",
+        )
+
+
+###
+### DB LAYER API
+###
+
+
+class FolderEntry(BaseModel):
+    id: _FolderID
+    parent_folder: _FolderID | None = Field(alias="traversal_parent_id")
+    name: str
+    description: str
+    owner: _GroupID = Field(alias="created_by")
+    created_at: datetime
+    last_modified: datetime
+    my_access_rights: _FolderPermissions
+    access_rights: dict[_GroupID, _FolderPermissions]
+
+    access_via_gid: _GroupID = Field(
+        ...,
+        description="used to compute my_access_rights, should be used by the frotned",
+    )
+    gid: _GroupID = Field(..., description="actual gid of this entry")
+
+    class Config:
+        orm_mode = True
+
+
+async def _get_top_most_access_rights_entry(
+    connection: SAConnection,
+    folder_id: _FolderID,
+    gid: _GroupID,
+    *,
+    permissions: _FolderPermissions | None,
+    enforece_all_permissions: bool,
+) -> RowProxy | None:
+
+    # Define the anchor CTE
+    access_rights_cte = (
+        sa.select(
+            [
+                folders_access_rights.c.folder_id,
+                folders_access_rights.c.gid,
+                folders_access_rights.c.traversal_parent_id,
+                folders_access_rights.c.original_parent_id,
+                folders_access_rights.c.read,
+                folders_access_rights.c.write,
+                folders_access_rights.c.delete,
+                sa.literal_column("0").label("level"),
+            ]
+        )
+        .where(folders_access_rights.c.folder_id == sa.bindparam("start_folder_id"))
```

Comment:
Looking already 5 mins into it and it is too complex for me :) I guess it somehow makes sense, but I would like to understand it, lets catch up in person please.

Reply:
use pydanbtic

Reply:
caught up with Matus.
added pedantic model